apps/labs/posts/sparse-array-ecosystem.md

```diff
@@ -0,0 +1,215 @@
+spx_mat_coo = scipy.sparse.coo_matrix(x)  #
+cpx_mat_csc = cupyx.sparse.csc_matrix(x)
+
+# Use `dok_array` or `dok_matrix` and then `.asformat`
+spy_arr_dok = scipy.sparse.dok_array(shape, dtype=...)  # Instantiate
+spy_arr_dok[...] = ...  # Assign repeatedly
+spy_arr_csr = spy_arr_dok.asformat("csr")  # Convert
+cpy_mat_csr = cupyx.sparse.csr_matrix(spy_arr_csr)  # Move to GPU
+
+# Use `array_type(tuple_of_constituent_arrays)`
+spz_arr_csr = scipy.sparse.csr_array((data, indices, indptr), shape=...)  # NumPy arrays
+cpz_mat_csr = cupyx.sparse.csr_array((data, indices, indptr), shape=...)  # CuPy arrays
+```
+
+The main pain points for this set of libraries are twofold:
+
+1. They only support 2-D arrays or matrices.
+2. Most of the existing work only exists for the deprecated `np.matrix` interface. SciPy has recently begun work to migrate away from `np.matrix`, but it remains the only option in `cupyx.sparse`.
+
+#### `torch.sparse`
+
+To create a sparse array, one can:
+
+```python
+# Call `.to_sparse or .to_sparse_* methods`
+spx = x.to_sparse([layout=...])
+spx_csr = x.to_sparse_csr()
+
+# Call constructors and use constituent arrays
+spy_coo = torch.sparse.coo_tensor(coords, data, [shape, dtype=...])
+spy_csr = torch.sparse.csr_tensor(indices, indptr, data, [shape, dtype=...])
+```
+
+The main pain points for `torch.sparse` is the lack of Array API support, differing API from most other libraries (such as lack of a `asformat` method, and format-specific methods), and lack of DOK support to construct hyper-sparse tensors intuitively. On the flip side, `torch.sparse` is the only library to allow for N-dimensional GPU sparse arrays.
+
+### Honorable Mentions
+
+There are a number of honorable mentions that haven't managed to gain enough traction to be included in this list:
+
+- [**`jax.experimental.sparse`**](https://jax.readthedocs.io/en/latest/jax.experimental.sparse.html): A forward-thinking sparse array module based on the [MLIR sparse tensor dialect](https://mlir.llvm.org/docs/Dialects/SparseTensorOps/).
+- [**`pysparse`**](https://pysparse.sourceforge.net/): Not to be confused with `sparse`, this was a package that didn't take off, but has a rather common name.
+- [**TACO**](http://tensor-compiler.org/): A C++ package with Python bindings by some of the brains behind performant sparse code generation.
+- [**`finch-tensor` and `Finch.jl`**](https://willowahrens.io/Finch.jl/dev): A Julia package with respective Python bindings that's also showing promise and contains some novel research on performant sparse computing.
+
+## Conclusion
+
+The Sparse array ecosystem in Python is, in a word, _fragmented_. The meme below illustrates the options available to users. We would like to change that for the better. To that end, we at Quansight along with Professor Samaringhe's group at MIT CSAIL have received a grant from DARPA spanning two years of work to help move the ecosystem in a direction that would benefit the whole commpunity.
```

Comment:
```suggestion
The Sparse array ecosystem in Python is, in a word, _fragmented_. The meme below illustrates the options available to users. We would like to change that for the better. To that end, we at Quansight, along with Professor Samaringhe's group at MIT CSAIL, have received a grant from DARPA spanning two years of work to help move the ecosystem in a direction that would benefit the whole community.
```

Reply:
There's no longer a "meme", correct?

Reply:
Also, yes, the ecosystem is fragmented, but also it is relatively nascent. It's a hard problem without a general solution. The DARPA grant is an attempt to provide a truly general solution.

Reply:
Done in https://github.com/Quansight/Quansight-website/pull/837/commits/4761fc3eb74286ade1d9344427b00551b44b5724.

> There's no longer a "meme", correct?

Removed the mention, thanks for catching that.

> Also, yes, the ecosystem is fragmented, but also it is relatively nascent. It's a hard problem without a general solution. The DARPA grant is an attempt to provide a truly general solution.

Added a few words to this effect.